### PR TITLE
Add support for enabling internal caching & bug fix

### PR DESCRIPTION
**Pull request by jcal-15**
* Add support for enabling internal caching to improve performance.
* Fixing a bug where fields of type Dict[...] were not being parsed properly.

### CODE REVIEW

**Comment by jcal-15** [comment]
I was comparing the performance of this package to another (much less robust) package I had. I noticed this was much slower. 
Which caching enabled it is much closer to the other package. 

That said, I don't really know if the performance matters too much in real world usage.

**Comment by jcal-15** [comment]
The performance improvement seems to be somewhere in the range of 3 - 4 times faster.

**Comment by jcal-15** [comment]
@wchresta, friendly reminder about this PR.

File: README.md [hunk 1]
```diff
@@ -100,3 +100,9 @@ for res in structured_response.results:
 
 ```
 
+## internal caching
```
Comment (wchresta) [inline README.md:103]:
```suggestion
## Internal caching
```

File: README.md [hunk 2]
```diff
@@ -100,3 +100,9 @@ for res in structured_response.results:
...
+
```
Comment (wchresta) [inline README.md:104]: How can use user activate caching?

File: from_dict/_caching.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
+import functools
+from typing import Optional
+
+from . import _from_dict
+
+_original_function = _from_dict._get_constructor_type_hints
+_cached_function: Optional['functools._lru_cache_wrapper'] = None
```
Comment (wchresta) [inline from_dict/_caching.py:7]: We must not use internals of packages we don't own. (The `_` in `_lru_cache_wrapper` marks it it an internal)

File: from_dict/_caching.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+def cache_enable(max_size: int = 100):
+    global _cached_function
+    if _cached_function:
+        if _cached_function.cache_info().maxsize == max_size:
+            return
+        _cached_function.cache_clear()
+    _cached_function = functools.lru_cache(max_size)(_original_function)
+    _from_dict._get_constructor_type_hints = _cached_function
```
Comment (wchresta) [inline from_dict/_caching.py:16]: I think we should make caching a first-class function of `_get_constructor_type_hints` instead of monkey-patching it in from the outside. This feels kind of wrong.

File: from_dict/_from_dict.py [hunk 5]
```diff
@@ -88,8 +88,19 @@ def get_constructor_type_hints(
     if cls is None:
         return None
 
-    return typing.get_type_hints(cls.__init__, global_ns, local_ns) or typing.get_type_hints(cls, global_ns, local_ns)
+    # Can't use caching if a namespace dictionary is passed in.
```
Comment (wchresta) [inline from_dict/_from_dict.py:91]: Why can't we? Is it because we're passing in `dict` and they cannot be hashed? Have you considered using `FrozenDict` from [frozendict](https://pypi.org/project/frozendict/) instead? They are hashable. So we could do something like:

```
@functools.lru_cache(max_size=1000)
def _get_constructor_type_hints(cls: Type, global_ns: Optional[FrozenDict], local_ns: Optional[FrozenDict]) -> Optional[Mapping[str, Type]]:
```

And then convert the incoming namespaces into frozendicts when we receive them.

File: ChangeLog.md [hunk 6]
```diff
@@ -1,3 +1,7 @@
+# Version 0.2.2 - 2022-06-21
+* Add support for enabling internal caching to improve performance.
+* Fixing a bug where fields of type Dict[...] were not being parsed properly.
```
Comment (wchresta) [inline ChangeLog.md:3]: Can we split this bugfix into its own commit?

File: from_dict/_caching.py [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+
+def cache_disable():
+    global _cached_function
+    if _cached_function:
+        _cached_function.cache_clear()
+        _cached_function = None
+    _from_dict._get_constructor_type_hints = _original_function
+
+def cache_is_enabled() -> bool:
```
Comment (wchresta) [inline from_dict/_caching.py:26]: I think there shouldn't be an option to enable and disable cache. We should be sure it works and then just have it always active. I feel the user shouldn't need to know about it and it's just an internal optimization we do to make it go supafast.
What would be a reason the user would want to disable this?

File: from_dict/_from_dict.py [hunk 8]
```diff
@@ -214,16 +225,18 @@ def handle_dict_argument(
...
+        if _get_constructor_type_hints(cls_arg_type_args[1]):
+            # Dict[a,b]; we only support b being a structure.
+            key_type, value_type = cls_arg_type_args
+            if fd_check_types:  # Perform type check on keys
+                all(type_check(k, key_type) for k in given_argument.keys())
+            argument_value = {
+                k: _from_dict(value_type, v)
+                for k, v in given_argument.items()
+            }
+        else:
```
Comment (wchresta) [inline from_dict/_from_dict.py:238]: Can we add a comment when this happens?

File: from_dict/_caching.py (hunk 7 above)
Comment (wchresta) [inline from_dict/_caching.py:26]: I think there shouldn't be an option to enable and disable cache. We should be sure it works and then just have it always active. I feel the user shouldn't need to know about it and it's just an internal optimization we do to make it go supafast.
What would be a reason the user would want to disable this?

File: from_dict/_from_dict.py (hunk 8 above)
Comment (wchresta) [inline from_dict/_from_dict.py:238]: Can we add a comment when this happens?

Comment (wchresta) [inline from_dict/_from_dict.py:238]: Can we add a comment when this happens?

File: from_dict/_caching.py (hunk 7 above)
Comment (wchresta) [inline from_dict/_caching.py:26]: I think there shouldn't be an option to enable and disable cache. We should be sure it works and then just have it always active. I feel the user shouldn't need to know about it and it's just an internal optimization we do to make it go supafast.
What would be a reason the user would want to disable this?

**Review by wchresta — CHANGES_REQUESTED** [review]
Great work, thank you very much and sorry for the late reply. I think I'm happy with merging a caching mechanism and also the bugfix. It would be great if we could make it such that there's no monkeypatching needed and the caching works for all cases, even if local_ns is given. Let me know if there's reasons why what wouldn't be better or not possible and we can discuss it :). In any case, thanks again.

**Comment by jcal-15** [comment]
Thanks for all the feedback. I should have explained why I made caching so complicated. I did not want to change the existing behavior in-case there were users who did not want have the extra memory overhead of caching. I am guessing the overhead is pretty small but I was not sure what others might think.

If this my own package I would have just enabled caching without the controls.

File: from_dict/_caching.py (hunk 3 above)
Comment (jcal-15) [reply from_dict/_caching.py:7]: Yes, I was concerned about this a bit too. I figured since it was just in a type hint it would not affect the package if it ever changed.
Since caching will always be enabled, this will not be an issue.

File: from_dict/_from_dict.py (hunk 5 above)
Comment (jcal-15) [reply from_dict/_from_dict.py:91]: I was think passing in global_ns and local_ns was such a small use case that it was not optimizing.

Whenever I passed in a value for global_ns or local_ns, I just called ```globals()``` or ```locals()```. Converting them to ```FrozenDict``` might undo an performance improvement that caching gives us and would require the client to have another dependency they don't want. I will investigate the performance impact. 

I will look into ways to make caching work with the namespace parameters. Feel free to over any other suggestions.

File: ChangeLog.md (hunk 6 above)
Comment (jcal-15) [reply ChangeLog.md:3]: Yes, I can do that.

File: from_dict/_caching.py (hunk 7 above)
Comment (jcal-15) [reply from_dict/_caching.py:26]: I think I was just overthinking the problem. 😄 

I was worried that you would not want caching at all and I thought if I disabled it by default it would easier to justify the change. It does add a lot of extra code to the package.

**Comment by jcal-15** [comment]
I think the best thing to do is to scrap this PR and start over.
